### PR TITLE
ccgx: Add quirk entries for the Goshen Ridge EVB

### PR DESCRIPTION
**Pull request by hughsie**
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [X] Feature
- [ ] Documentation

### CODE REVIEW

**Comment by hughsie** [comment]
Ryan, do you know if this requires a `CcgxDmcTriggerCode`?

**Comment by hughsie** [comment]
@bleungatchromium FYI

**Comment by hughsie** [comment]
> it is not a mass product and it is just a reference to make a vendor product using Intel Goshen Ridge, DMC and PD controllers

Agree.

> I have not gotten any requests to support Gatkex Creek in LVFS.

Google asked me to have a look, I believe on the logic that most USB4 docks are going to be based on Goshen Ridge and will probably also be very close to the reference design. All of the USB4 docks that want the WWCB certification will need fwupd support, and so it made sense to ensure the reference design is well supported.

> Gatkex Creek needs CcgxDmcTriggerCode = 1,

Thanks, that's what I wanted to know! :)

> Also need to InstallDuration and RemoveDelay but not measured by my side.

Out of interest, do you have any firmware I can use for Gatkex Creek so I can measure it? Thanks.

File: plugins/ccgx/ccgx.quirk [hunk 1]
```diff
@@ -112,3 +112,9 @@ CcgxDmcTriggerCode = 0x01
 CcgxImageKind = dmc-composite
 InstallDuration = 898
 RemoveDelay = 732000
+
+# Intel Gatkex Creek Goshen Ridge EVB
+[USB\VID_04B4&PID_5220]
```
Comment (superm1) [inline plugins/ccgx/ccgx.quirk:117]: Are we sure that this exact PID is unique to Gatkex creek? If not - it should go into a noinst ccgx quirk.

Comment (hughsie) [reply plugins/ccgx/ccgx.quirk:117]: @chlee75 please

**Comment by bleungatchromium** [comment]
Gatkex is a good start, and I think a few developers on my side have one, but more common ones in the mass market are based on Intel's dock references (which are basically white boxed into actual products).

These were:
Godzilla Creek
Gorilla Creek

Just as an example, here is the Godzilla Creek reference board I have with a Quanta VID:
```
[75606.785233] usb 3-1.5: new full-speed USB device number 48 using xhci_hcd
[75606.881574] usb 3-1.5: New USB device found, idVendor=2bef, idProduct=9065, bcdDevice= 0.00
[75606.881585] usb 3-1.5: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[75606.881589] usb 3-1.5: Product: QSI Thunderbolt4 Godzilla Hub
[75606.881591] usb 3-1.5: Manufacturer: Quanta Storage Inc.
[75606.881594] usb 3-1.5: SerialNumber: 11AD1D0AB6073C08071A0B00
```

**Comment by hughsie** [comment]
@bleungatchromium is the usb4 parent of 2bef:9065 still 8087:0B40?

@chlee75 can you share the CcgxDmcTriggerCode parameters for Godzilla Creek and Gorilla Creek too please?

**Comment by bleungatchromium** [comment]
> @bleungatchromium is the usb4 parent of 2bef:9065 still 8087:0B40?
> 


Here's the USB 3 & 2 tree.

```
[75606.207532] usb 2-1: new SuperSpeedPlus Gen 2 USB device number 27 using xhci_hcd
[75606.220841] usb 2-1: New USB device found, idVendor=8087, idProduct=0b40, bcdDevice=12.34
[75606.220859] usb 2-1: New USB device strings: Mfr=1, Product=2, SerialNumber=0
[75606.220865] usb 2-1: Product: USB3.0 Hub
[75606.220870] usb 2-1: Manufacturer: Intel Corporation.
[75606.223455] hub 2-1:1.0: USB hub found
[75606.223641] hub 2-1:1.0: 4 ports detected
[75606.239054] usb 3-1: New USB device found, idVendor=1d5c, idProduct=5801, bcdDevice= 1.01
[75606.239068] usb 3-1: New USB device strings: Mfr=1, Product=2, SerialNumber=0
[75606.239074] usb 3-1: Product: USB2.0 Hub
[75606.239078] usb 3-1: Manufacturer: Fresco Logic, Inc.
[75606.241945] hub 3-1:1.0: USB hub found
[75606.241990] hub 3-1:1.0: 5 ports detected
[75606.785233] usb 3-1.5: new full-speed USB device number 48 using xhci_hcd
[75606.881574] usb 3-1.5: New USB device found, idVendor=2bef, idProduct=9065, bcdDevice= 0.00
[75606.881585] usb 3-1.5: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[75606.881589] usb 3-1.5: Product: QSI Thunderbolt4 Godzilla Hub
[75606.881591] usb 3-1.5: Manufacturer: Quanta Storage Inc.
[75606.881594] usb 3-1.5: SerialNumber: 11AD1D0AB6073C08071A0B00
```


2bef:9065 is hanging off of a USB 2.0 hub from Fresco.
Intel's USB3 hub exists though with the vid:pid you mentioned. 8087:0b40, but it's not connected to the billboard device.

**Comment by chlee75** [comment]
@bleungatchromium 
>Gatkex is a good start, and I think a few developers on my side have one, but more common ones in the mass market are based on Intel's dock references (which are basically white boxed into actual products).

Hi. Benson. 
I am an Infineon USB-C engineer and could you help me on my questions.

Does Google have a plan to support Intel Goshen Ridge EVBs in Chromebook through FWUPD?
Intel Goshen Ridge EVBs are also Intel's properties.
And Do you have any discussion with Intel about this?

**Comment by chlee75** [comment]
@hughsie 
>can you share the CcgxDmcTriggerCode parameters for Godzilla Creek and Gorilla Creek too please?

I need to discuss with my company dock team leader for Intel Goshen Ridge EVB information sharing such as firmware binaries, USB connection structure etc.
I will let you know after the discussion.

**Comment by hughsie** [comment]
> clear intention for Goshen Ridge EVBs support in LVFS

I don't necessarily think we need or want packages on the LVFS for the EVB. I do think we want to make it easy for vendors who are using one of the EVBs as a base for _their_ dock design that *does* need LVFS support. @bleungatchromium can explain the Google stance better than me perhaps.

**Comment by bleungatchromium** [comment]
Hi @chlee75 and @hughsie 

Ok, from the WWCB point of view, our priority are the Intel reference platforms that are closer to white-box.

From my knowledge, the two Goshen Ridge references that are closest to consumer products are:
Godzilla Creek
Gorilla Creek

The both use Infineon CCG for PD controllers, and have a DMC.

For development, my team does have Gatkex Creek boards as well, but updating these are not as high a priority for us as the other two reference boards.

Upgrading the Gatkex Creek would be great for development purposes, but it is not as critical as upgrading the released docks that our customers actually have, which will be based on Godzilla and Gorilla.

Yes, my team has also been engaging with Intel on this project.

In summary, I think it's fine that Gatkex is supported, but the priority should be on the reference platforms that are closer to consumer versions.

By the way, I also opened issue #4803 for Godzilla Creek support, which I hope will just take the place of an extra quirk entry as well.

Thanks,
Benson

**Comment by chlee75** [comment]
@bleungatchromium , @hughsie 

1. About the Goshen Ridge EVB support in LVFS such as Godzilla, Gorilla and Gatkex Creek etc.,
The EVBs use Infineon USB VID/PID in DMC IC.
And the DMC USB VID/PID is not unique in each EVB because it is not a mass product but just a reference design.
If the same DMC part is used, the DMC VID/PID will be the same in the EVB.
For example, Gatkex and Godzilla Creek have the same USB VID/PID in DMC while Gorilla Creek has a different one.
So I think that it is impossible to support Goshen Ridge EVBs in LVFS because DMC USB VID/PIDs are not unique.
However, customer dock based on the EVB can be supported in LVFS because customers can assign their DMC VID/PID to the their docks with an Infineon utility.
  
2. About Customer Goshen Ridge Dock support in LVFS and Chromebook, 
Hi. Benson.
It seems that you have customer products based on Godzilla and Gorilla.
Please contact Satyajit (Satyajit.Sarma@infineon.com) who is a docking team leader in Infineon for any dock support in both LVFS and Chromebook.
Hardware and firmware features could be different according to customer dock design even though the customer dock is based on the EVBs.
So you need Infineon docking team support for the customer product.

**Comment by hughsie** [comment]
> So I think that it is impossible to support Goshen Ridge EVBs in LVFS because DMC USB VID/PIDs are not unique.

Agree; lets just do fwupd support so the hardware appears and *not* upload firmware to the LVFS.

> Trigger Code, Installation Time, and Remove Delay may be different.

It's easy enough to shared VID/PIDs and trigger codes, right? There's nothing secret there?

**Comment by superm1** [comment]
> However, customer dock based on the EVB can be supported in LVFS because customers can assign their DMC VID/PID to the their docks with an Infineon utility.

I picked up on this bit specifically - and it tells me we really want to keep the EVB VID/PID as noinst to ensure that you never have a situation that VID/PID can be used across docks.

@bleungatchromium is there anything in the WWCB program that already enforces this?  Some sort of VID/PID disallow list?  If not, I would encourage you to add that.

**Comment by hughsie** [comment]
> it tells me we really want to keep the EVB VID/PID as noinst

Ack; agree. Updated.

**Review by superm1 — APPROVED** [review]

**Comment by chlee75** [comment]
@hughsie , @superm1 , @bleungatchromium 

>Agree; lets just do fwupd support so the hardware appears and not upload firmware to the LVFS.

Gatkex and Godzilla Creek use CY7C65219 and Gorilla Creek use CYUSB4357 DMC IC.
All EVB that use CY7C65219 will have the same USB VID/PID(04B4/5220) and All EVB that use CYUSB4357 will have the same USB VID/PID(04B4/521B).
It means if the next generation Intel Ridge EVB uses CY7C65219 then, it will have the same VID/PID of Goshen Ridge EVB (Gatkex and Godzilla). 
So the next generation Intel Ridge EVB will be detected as Goshen Ridge EVB.
In addition, Gatkex and Godzilla will be detected as the same device because it has the same VID/PID.
It is not the result that we expect.

So I think that Intel Ridge EVB support in LVFS is not proper.

Ryan.

**Comment by superm1** [comment]
Right; so we're in agreement with you.  No LVFS support, but the PID/VID is put as a "noinst" quirk.  This means that it's in source tree only, not installed on target system.  If you want to do development and experiments with EVB you can copy it to a system, but this would never be in production and can't be uploaded to LVFS.

**Comment by chlee75** [comment]
@superm1 
>so we're in agreement with you. No LVFS support, but the PID/VID is put as a "noinst" quirk. This means that it's in source tree only, not installed on target system

The USB\VID_04B4&PID_5220 is the VID/PID of CY7C65219 silicon and it is not used only for Intel Gatkex Creek Goshen Ridge.
It can be used any EVB that uses CY7C65219.
So the quirk information that USB\VID_04B4&PID_5220 indicates Intel Gatkex Creek Goshen Ridge is incorrect.

That's why I don't want to add Intel EVB into the quirk file.

Ryan.

**Comment by hughsie** [comment]
>  So the quirk information that USB\VID_04B4&PID_5220 indicates Intel Gatkex Creek Goshen Ridge is incorrect.

**Comment by chlee75** [comment]
@hughsie 

> is CcgxDmcTriggerCode going to be try for all CY7C65219 and CYUSB4357 EVB products?

By default, trigger code is 1 (update and reboot within remove delay) and it is used only to update the firmware. 
In the future, I am not sure about the trigger code. 
But think that trigger code 1 is OK because the EVBs will not support firmware update.

RemoveDelay and InstallDuration are also used only when to update the firmeware.
In Gatkex,  InstallDuration is 59 sec and RemoveDelay  is 48000 msec. 
So I think InstallDuration (72 sec with +25% more) and RemoveDelay (60000 msec with +25%) would be OK.
I did not measured RemoveDelay and InstallDuration in CYUSB4357. I expect that it will be similar to CY7C65219.
But those InstallDuration and RemoveDelay value are OK because the EVBs will not support firmware update.

CcgxImageKind = dmc-composite is needed.

So is your intention to show an example to developers for the Infineon DMC plugin?

**Comment by hughsie** [comment]
> CcgxImageKind = dmc-composite

For both, or just CYUSB4357? Thanks.

**Comment by chlee75** [comment]
@hughsie 

>For both, or just CYUSB4357? Thanks.
CcgxImageKind = dmc-composite for both.

Is "Name" field not necessary?

EVBs do not support firmware updates and it just shows an example.  
I am not sure that it is worth adding EVB into the qurik.

@bleungatchromium 
Is it OK for Google?
